Kernel/FileSystem/InodeWatcher.h

```diff
@@ -1,44 +1,48 @@
 /*
  * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2021, the SerenityOS developers.
  *
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
 #pragma once
 
 #include <AK/Badge.h>
-#include <AK/CircularQueue.h>
-#include <AK/WeakPtr.h>
-#include <Kernel/API/InodeWatcherEvent.h>
-#include <Kernel/FileSystem/File.h>
-#include <Kernel/Lock.h>
+#include <AK/HashMap.h>
+#include <AK/NonnullRefPtr.h>
+#include <Kernel/FileSystem/WatchDescriptor.h>
+#include <Kernel/FileSystem/WatchFile.h>
 
 namespace Kernel {
 
 class Inode;
 
-class InodeWatcher final : public File {
+class InodeWatcher final : public WatchFile {
 public:
-    static NonnullRefPtr<InodeWatcher> create(Inode&);
+    static NonnullRefPtr<InodeWatcher> create();
     virtual ~InodeWatcher() override;
 
-    virtual bool can_read(const FileDescription&, size_t) const override;
-    virtual bool can_write(const FileDescription&, size_t) const override;
-    virtual KResultOr<size_t> read(FileDescription&, u64, UserOrKernelBuffer&, size_t) override;
-    virtual KResultOr<size_t> write(FileDescription&, u64, const UserOrKernelBuffer&, size_t) override;
+    virtual KResult close() override;
+
     virtual String absolute_path(const FileDescription&) const override;
     virtual const char* class_name() const override { return "InodeWatcher"; };
 
-    void notify_inode_event(Badge<Inode>, InodeWatcherEvent::Type);
-    void notify_child_added(Badge<Inode>, const InodeIdentifier& child_id);
-    void notify_child_removed(Badge<Inode>, const InodeIdentifier& child_id);
+    virtual bool is_iwatcher() const override { return true; }
+
+    void notify_inode_event(Badge<Inode>, InodeIdentifier, InodeWatcherEvent::Type, StringView const& name = {});
+
+    KResultOr<NonnullRefPtr<WatchDescriptor>> register_inode(unsigned fd, Inode&, unsigned event_mask);
+    KResult unregister_by_fd(unsigned fd);
+    void unregister_by_inode(Badge<Inode>, InodeIdentifier);
 
 private:
-    explicit InodeWatcher(Inode&);
+    // FIXME(before merge): do we need this or ::create above?
```

Comment:
Ditto this, should I keep the `::create` pattern as with other `File` subclasses or just get rid of it and instantiate directly?